design/ironic-standalone-operator.md

```diff
@@ -0,0 +1,369 @@
+   service.
+2. Due to the way Ironic is designed, each replica will handle its proportion
+   (1/3 in most cases) of nodes (active/active architecture, not
+   active/backup).
+
+MariaDB is not going to be run in an HA fashion. The mid-term plans include
+looking into using a persistent volume for it instead.
+
+When the HA architecture is enabled via a flag on the `Ironic` resource, all
+Ironic components (except for MariaDB and dnsmasq) will be installed in a
+*DaemonSet* instead of a *Deployment*.
+
+##### Dnsmasq, iPXE and provisioning network
+
+Dnsmasq is also not going to be run with more than 1 replica. It's not
+impossible to run several DHCP servers on the same network, but it's harder to
+configure and to debug. In the future, we might look into some sort of a
+managed DHCP offering, e.g. [Kea][kea].
+
+Using a provisioning network will require having a provisioning IP per each
+control plane node instead of only one with the non-HA architecture.
+
+Using iPXE in the HA configuration poses one more problem. Our (static) DHCP
+configuration must point each host at its iPXE configuration script. However,
+dnsmasq does not know, which host belongs to which Ironic instance. To tackle
+this limitation, a new [boot configuration API][boot config] has been proposed
+(but not yet implemented) in Ironic. It will allow our DHCP configuration to
+always point at the same Ironic instance for iPXE configuration, and Ironic
+itself will do the required routing.
+
+[boot config]: https://specs.openstack.org/openstack/ironic-specs/specs/approved/boot-config-api.html
+
+##### JSON RPC
+
+Ironic itself is a clustered software. Each instance, as noted above, will
+handle its share of all nodes. When an instance crashes, the remaining
+instances will take over its responsibilities. You can hit the API on any
+instance for any node, and the request will be forwarded to the right instance.
+
+To achieve that, Ironic supports JSON RPC. Metal3 currently does not use it,
+and it still will not be used in the non-HA case. For JSON RPC to be usable,
+each Ironic instance must register its RPC access IP or hostname in the
+database.
+
+When TLS support is enabled, the RPC communication must be secured by
+TLS as well. This may pose a problem since each Ironic instance needs a TLS
+certificate that is valid for its RPC access IP or hostname. This problem has
+been extensively discussed in the [initial HA proposal][issue3], and here is
+the proposed solution, at least for the MVP case:
+
+The Ironic controller will generate a self-signed CA and pass its public and
+private parts into each Ironic container. Each Ironic container will generate
+its private key certificate and sign the certificate with this CA. The CA will
+be trusted **only** for the RPC purpose, removing the possibility of abuse.
+To reduce the number of code paths, this process will happen unconditionally,
+even when TLS for Ironic itself is not enabled.
```

Comment:
Just for my understanding what prevents us from enabling TLS unconditionally for ironic itself?

Reply:
Well, nothing really. We'd need to decide if we want to use a CertificateSigningRequest (and require a manual approval) or just generate a self-signed certificate. Opinions?

Reply:
I guess the end goal would be to have an integration with cert-manager as in most K8s environments cert-manager is handling the certificates.

Reply:
My only worry about the manual step (approving the cert request). Is it also considered standard?

I mean, in the situation when the admin has not provided any certificate via the secret.